hoot-core/src/main/cpp/hoot/core/algorithms/WayJoiner.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2018 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#ifndef WAYJOINER_H
+#define WAYJOINER_H
+
+// TGS
+#include <tgs/SharedPtr.h>
+
+// Hoot
+#include <hoot/core/OsmMap.h>
+#include <hoot/core/elements/Way.h>
+
+namespace hoot
+{
+
+class WayJoiner
+{
+public:
+  WayJoiner(const OsmMapPtr& map);
+
+  /**
+   * @brief join Runs all joining algorithms
+   */
+  static void join(const OsmMapPtr& map);
+  void join();
+
+private:
+
+  /**
+   * @brief joinParentChild Simplest joining algorithm that joins a way with a parent id to that parent
+   */
+  void joinParentChild();
+
+  /**
+   * @brief joinSiblings Joining algorithm that searches for all ways that have the same parent id,
+   *    attempts to order them into adjoining way order, then joins them
+   */
+  void joinSiblings();
+
+  /**
+   * @brief joinAtNode Joining algorithm that searches all ways that have a parent id and tries
+   *    to join them with adjacent ways that have the same tags
+   */
+  void joinAtNode();
+
+  /**
+   * @brief areJoinable Check the status of the ways to see if they are compatible when joining at a node
+   *    essentially UNKNOWN1 and UNKNOWN2 ways aren't joinable together
+   */
+  bool areJoinable(const WayPtr& w1, const WayPtr& w2);
+
+  /**
+   * @brief resetParents Resets parent id for all ways after joining operation has completed
```

Comment:
Add a comment here that this will do nothing if _leavePid is true. ALSO, there should probably be public functions for setting/getting _leavePid.